File: reference/python/aicp_ref/chain.py
from __future__ import annotations

from typing import Any


def verify_transcript_chain(messages: list[dict[str, Any]]) -> list[str]:
    errors: list[str] = []
    prev_hash = None
    for idx, msg in enumerate(messages, start=1):
        current_prev = msg.get("prev_msg_hash")
        if prev_hash is not None:
            if current_prev is None:
                errors.append(f"line {idx}: missing prev_msg_hash for non-first message")
            elif current_prev != prev_hash:
                errors.append(
                    f"line {idx}: prev_msg_hash mismatch (expected {prev_hash}, got {current_prev})"
                )
        prev_hash = msg.get("message_hash")
    return errors

File: sdk/typescript/src/chain.js
export function verifyPrevHashChain(messages) {
  const errors = [];
  let prev;
  for (let i = 0; i < messages.length; i += 1) {
    const msg = messages[i];
    if (prev !== undefined && msg.prev_msg_hash !== undefined && msg.prev_msg_hash !== prev) {
      errors.push(`line ${i + 1}: prev_msg_hash mismatch (expected ${prev}, got ${msg.prev_msg_hash})`);
    }
    prev = msg.message_hash;
  }
  return errors;
}
